web/src/compose_actions.ts

```diff
@@ -516,8 +517,28 @@ export function on_narrow(opts: NarrowActivateOpts): void {
             }
             return;
         }
-        // Do not open compose box if organization has disabled sending
-        // direct messages and recipient is not a bot.
+        // Do not open compose box if sender is not allowed to send direct message.
+        const recipient_ids_string = opts.private_message_recipient
+            .split(",")
+            .map((email) => people.get_by_email(email)?.user_id)
+            .join(",");
+
+        const previous_messages_exist = all_messages_data
+            .all_messages()
+            .find((message) => message.is_private && message.to_user_ids === recipient_ids_string);
```

Comment:
> Can we move this check to people.user_can_initiate_direct_message_thread? I think it would be helpful to avoid duplication.

Doing that would require importing `all_messages_data.ts` which results in a dependency cycle. So it can't be moved to `people.user_can_initiate_direct_message_thread`. I have refactored some code and added `previous_direct_messages_exist` function in `message_util.ts`. I am not sure if that is a good place to add this helper function. What do you think? Maybe we can make a separate file for all the helper functions related to this setting?

Reply:
> will this check work as expected if there exists a direct message for the set of users but the data is not available with the client.

As it checks `all_messages_data` for previous messages, it will not allow the user to send messages if the messages are not available locally. I guess this will result in unexpected behaviour in 2 cases:
- Visibility of `Send direct message` button on the user card popover.
- When manually adding user pills in the compose box and the view is not narrowed.

Do I need to add a separate function to fix this?

Reply:
I guess `message_util` is a fine place to add a helper function. And for the import cycle, we have similar functions which checks the permission in `settings_data.ts`, so if moving the functions from `people.ts` to `settings_data.ts` can help in doing the check for whether previous message exists in the functions which checks permission.

Reply:
> As it checks all_messages_data for previous messages, it will not allow the user to send messages if the messages are not available locally. 

That will be a problem. It will be fine to show UI/button when user is not allowed to send messages because server will return an error, but not the other way around.

Solution other than just showing the UI to send messages is check by sending a request to server. Might be good to discuss on CZO by starting a thread in #frontend.

Reply:
> moving the functions from people.ts to settings_data.ts can help in doing the check for whether previous message exists in the functions which checks permission.

Moving it to `settings_data.ts` requires importing `people.ts` in `settings_data.ts` - this creates an import cycle too.

Reply:
> Might be good to discuss on CZO by starting a thread in #frontend.

Started a discussion: https://chat.zulip.org/#narrow/stream/6-frontend/topic/new.20dm.20settings.20UI/near/1839837

Reply:
This discussion led to #30659. Please cross-link in all places when opening a PR for something discussed in another.